File: ALDS1/3_D_AreasOnTheCrossSectionDiagram.c
#include <stdio.h>
#include <stdlib.h>

// 水たまり構造体
typedef struct {
    int leftSideIndex; // 左端のインデックス
    int area; // 面積
} Puddle;

int main() {
    int leftSideStack[20000]; // '\'が出現したインデックスを保持するスタック
    int leftSideStackTop = 0;
    Puddle puddles[10000]; // 水たまりのスタック
    int puddlesTop = 0;
    char inStr[20000];
    int areaSum = 0;

    scanf("%s", inStr);

    for (int i = 0;; i++) {
        if (inStr[i] == '\\') {
            leftSideStack[leftSideStackTop++] = i;
        }
        else if (inStr[i] == '/' && leftSideStackTop > 0) {
            int leftSideIndex = leftSideStack[--leftSideStackTop]; // 直近の'\'のインデックスを取得する
            int area = i - leftSideIndex;
            areaSum += area;

            // 水たまりの統合を行う
            //     現在処理中の水たまりの左端より右側にある水たまりは統合を行う
            while (puddlesTop && puddles[puddlesTop - 1].leftSideIndex > leftSideIndex) {
                area += puddles[--puddlesTop].area;
            }
            puddles[puddlesTop].leftSideIndex = leftSideIndex;
            puddles[puddlesTop++].area = area;
        }
        else if (inStr[i] == '\0') break;
    }

    // 結果を出力
    printf("%d\n", areaSum);
    printf("%d", puddlesTop);
    for (int i = 0; i < puddlesTop; i++) {
        printf(" %d", puddles[i].area);
    }
    printf("\n");

    return 0;
}

File: ALDS1/3_D_AreasOnTheCrossSectionDiagram.py
# 水たまりクラス
#     self.leftSideIndex: 水たまりの左端のインデックス
#     self.area: 水たまりの面積
class Puddle:
    def __init__(self, leftSideIndex: int, area: int):
        self.leftSideIndex: int = leftSideIndex
        self.area: int = area

leftSideStack = [] # '\'が出現したインデックスを保持するスタック
puddles = [] # 水たまりのリスト

for i, c in enumerate(input()):
    if c == '\\':
        leftSideStack.append(i)
    if c == '/' and len(leftSideStack) > 0:
        leftSideIndex = leftSideStack.pop() # 直近の'\'のインデックスを取得する
        area = i - leftSideIndex
        
        # 水たまりの統合を行う
        #     現在処理中の水たまりの左端より右側にある水たまりは統合を行う
        while puddles and puddles[-1].leftSideIndex > leftSideIndex:
            area += puddles.pop().area
        
        puddles.append(Puddle(leftSideIndex, area))

# 結果を出力
print(sum([i.area for i in puddles]))
print(len(puddles), end='')
if puddles:
    print(' ' + ' '.join([str(i.area) for i in puddles]), end='')
print()
